tools/autograd/templates/Functions.cpp

```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
   return u_term + sigma_term + v_term;
 }
 
+Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
+                    bool eigenvectors, const Tensor& lambda, const Tensor& v) {
+  // This gradient only works for real eigenvalues at the moment.
+  TORCH_CHECK(eigenvectors,
+           "eig_backward: Setting eigenvectors to false in torch.eig doesn't compute eigenvectors ",
+           "and hence we cannot compute backward. Please use torch.eig(eigenvectors=True)");
+  auto imag_idx = at::ones({1}, lambda.options().dtype(at::kLong));
+  auto real_idx = at::zeros({1}, lambda.options().dtype(at::kLong));
+  auto zeros = at::zeros({1}, lambda.options());
+  TORCH_CHECK(
+      at::allclose(
+        at::index_select(lambda, /*dim=*/-1, /*index=*/imag_idx),
+        zeros),
+      "eig_backward: Backward calculation does not support complex eigenvalues at the moment.");
+
+  auto glambda = grads[0];
+  auto gv = grads[1];
+  auto vt = v.transpose(-2, -1);
+
+  Tensor result;
+  // contribution from the eigenvectors
+  if (gv.defined()) {
+    auto rlambda = at::index_select(lambda, /*dim=*/-1, /*index=*/real_idx);
+
+    auto hm = rlambda.transpose(-2,-1) - rlambda;
+    hm.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+    hm.pow_(-1.0);
+
+    auto gvortho = gv - at::sum(gv * v, /*dim=*/-2, /*keepdim=*/true) * v;
+    auto B = hm * at::matmul(vt, gvortho);
+    auto A = at::matmul(B, vt);
+
+    std::tie(result, std::ignore) = at::solve(A, vt);
+  }
+  else {
+    result = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  }
+  // contribution from eigenvalues
+  if (glambda.defined()) {
+    Tensor result1;
+    auto grlambda = at::index_select(glambda, /*dim=*/-1, /*index=*/real_idx) * vt;
+    auto A = at::matmul(v, grlambda);
+    auto vvt = at::matmul(v, vt);
+    std::tie(result1, std::ignore) = at::solve(A, vvt);
+    result = result.add(result1);
```

Comment:
This will only work if result is already defined. You want to protect it with `if (result.defined())` and otherwise just do the assignment.

Reply:
The result has already been defined in the previous block. If `gv.defined()`, then result is defined as one of the output of `solve()`. Otherwise, it is defined as `zeros_like(self)`.

The idea comes from `symeig_backward` which uses the similar strategy.

Reply:
Ho right. But then the `else` part of the previous block is not needed. It create a Tensor full of zeros that is not necessary :)

Reply:
Yes, you're right. I've changed it in the new commit.